book/6-export.md

```diff
@@ -0,0 +1,289 @@
+  text_representation:
+    extension: .md
+    format_name: myst
+    format_version: 0.13
+    jupytext_version: 1.16.4
+kernelspec:
+  display_name: Python 3 (ipykernel)
+  language: python
+  name: python3
+---
+
+# Converting sub-volumes to/from other file formats
+
++++
+
+In this chapter we'll look at how to convert sub-volumes of 3D zarr images to and from other file formats.
+The use case for this is allow users to extract sub-volumes of large 3D images into a preferred file format,
+perform some analysis/processing on the sub-volume using their preferred tools and then
+copy the results back in-place to the original zarr image.
+
+Note: this chapter (and the whole of the textbook) uses Zarr version 2 and OME-Zarr version 0.4. Although newer versions exist (Zarr version 3 and OME-Zarr version 0.5), software tools have not quite caught up to support the newer versions.
+
+```{code-cell} ipython3
+import napari
+
+import nibabel as nib
+import pydicom
+from pydicom.dataset import Dataset, FileDataset
+import datetime
+```
+
+## Converting Zarr sub-volumes to tiff
+
+To begin this chapter, we'll look at how to convert Zarr sub-volumes to tiff.
+
+To start with we'll load our heart sample dataset, as a zarr array
+and then extract a smaller volume as a NumPy array.
+
+Note: the heart sample dataset is only 200MB,
+so the entire dataset could be extracted into a NumPy array (and therefore loaded into memory) if required.
+If you are working with larger datasets, you will need to keep the extracted volume small enough to fit into memory, or extract it as a series of smaller chunks.
+
+```{code-cell} ipython3
+from pathlib import Path
+import zarr
+import numpy as np
+
+data_path: data_path = (Path(__file__).parent / "data" / "hoa_heart.zarr").resolve()
+zarr_array = zarr.open_array(data_path)
+np_array = zarr_array[100:200, 100:200, 100:200]
+n_bytes = 100*100*100*zarr_array.dtype.itemsize  
+n_gb = n_bytes / (10**9)
+```
+
+The next step is to write this NumPy array to a tiff file.
+We'll use the `imageio` library to do this.
+
+Processing of the sub-volume can be done using any tool
+that can read and write tiff files.
+For example, you could use `Napari`.
+
+The `imageio` library is used to then read the
+tiff file back into a NumPy array.
+This subvolume array can then be copied back
+in place to the original Zarr array.
+
+```{code-cell} ipython3
+import imageio.v3 as iio
+iio.imwrite("image_file.tiff", np_array, plugin='tifffile')
+
+# Do some processing with another tool / software here...
+# Then load the result:
+zarr_small=iio.imread("image_file.tiff", plugin='tifffile')
+
+zarr_array[100:200, 100:200, 100:200]=zarr_small
+
+
+```
+
+To view the results, we can use the `napari` viewer
+to display the Zarr array including the newly written sub-volume.
+
+```{code-cell} ipython3
+viewer = napari.Viewer()
+new_layer = viewer.add_image(zarr_array)
+napari.run()
+```
+
++++
+
+## Converting Zarr sub-volumes to NIfTI/DICOM
```

Comment:
Given the concept of saving/loading is the same (export to numpy array, save numpy array, process, load numpy array, insert into Zarr array), I'm not sure if there's much benefit in extended examples after here. Perhaps what would be nice is a reference table that links to examples in other libraries that show how to save/load numpy data. That way we don't duplicate example material here, and it saves us effort if we want to point to other data formats.